auslib/db.py

```diff
@@ -1115,6 +1118,51 @@ def mergeUpdate(self, old_row, what, changed_by, transaction=None):
             self.log.debug("Merged %s into scheduled change '%s'", what, sc["sc_id"])
 
 
+class SignoffsTable(AUSTable):
+
+    def __init__(self, db, metadata, dialect, baseName):
+        self.table = Table("{}_signoffs".format(baseName), metadata,
+                           Column("sc_id", Integer, primary_key=True, autoincrement=False),
+                           Column("username", String(100), primary_key=True),
+                           Column("role", String(50), nullable=False),
+                           )
+        # Because Signoffs cannot be modified, there's no possibility of an
+        # update race, so they do not need to be versioned.
+        super(SignoffsTable, self).__init__(db, dialect, versioned=False)
+
+    def insert(self, changed_by=None, transaction=None, dryrun=False, **columns):
+        if "sc_id" not in columns or "role" not in columns:
+            raise ValueError("sc_id and role must be provided when signing off")
+        if "username" in columns and columns["username"] != changed_by:
+            raise PermissionDeniedError("Cannot signoff on behalf of another user")
+        if not self.db.hasRole(changed_by, columns["role"], transaction=transaction):
+            raise PermissionDeniedError("{} cannot signoff with role '{}'".format(changed_by, columns["role"]))
+
+        existing_signoff = self.select({"sc_id": columns["sc_id"], "username": changed_by}, transaction)
+        if existing_signoff:
+            # It shouldn't be possible for there to be more than one signoff,
+            # so not iterating over this should be fine.
+            existing_signoff = existing_signoff[0]
+            if existing_signoff["role"] != columns["role"]:
+                raise PermissionDeniedError("Cannot signoff with a second role")
+            # Signoff already made under the same role, we don't need to do
+            # anything!
+            return True
```

Comment:
sqlalchemy n00b question: What does `return True` allows, compared to just simply `return`?